File: ocaml/lib/day1.ml
let p1 input =
  let lines =
    String.split_on_char '\n' input
    |> List.filter (fun line ->
      match line with
      | "" -> false
      | _ -> true)
  in
  let res =
    List.fold_left
      (fun acc l ->
        let chars = l |> String.to_seq |> List.of_seq in
        let first = List.find (fun ch -> '0' <= ch && ch <= '9') chars in
        let last =
          List.rev chars |> List.find (fun ch -> '0' <= ch && ch <= '9')
        in
        let combined = String.make 1 first ^ String.make 1 last in
        let num = int_of_string combined in
        acc + num)
      0
      lines
  in
  Printf.printf "%d\n" res
;;

let replace str old_sub new_sub =
  let re = Re.Posix.compile_pat old_sub in
  Re.replace_string re ~by:new_sub str
;;

let p2 input =
  let lines =
    String.split_on_char '\n' input
    |> List.filter (fun line ->
      match line with
      | "" -> false
      | _ -> true)
  in
  let res =
    List.fold_left
      (fun acc l ->
        let s1 = replace l "one" "one1one" in
        let s2 = replace s1 "two" "two2two" in
        let s3 = replace s2 "three" "three3three" in
        let s4 = replace s3 "four" "four4four" in
        let s5 = replace s4 "five" "five5five" in
        let s6 = replace s5 "six" "six6six" in
        let s7 = replace s6 "seven" "seven7seven" in
        let s8 = replace s7 "eight" "eight8eight" in
        let s = replace s8 "nine" "nine9nine" in
        let chars = s |> String.to_seq |> List.of_seq in
        let first = List.find (fun ch -> '0' <= ch && ch <= '9') chars in
        let last =
          List.rev chars |> List.find (fun ch -> '0' <= ch && ch <= '9')
        in
        let combined = String.make 1 first ^ String.make 1 last in
        let num = int_of_string combined in
        acc + num)
      0
      lines
  in
  Printf.printf "%d\n" res;
  ()
;;

File: rs/src/day1.rs
pub fn part_one(input: &str) -> i32 {
    let t = input.lines().into_iter().fold(0, |acc, l| {
        let first = l.chars().find(|ch| '0' <= *ch && *ch <= '9').unwrap();
        let last = l.chars().rev().find(|ch| '0' <= *ch && *ch <= '9').unwrap();
        let t = format!("{}{}", first, last);
        let res: i32 = t.parse().unwrap();
        acc + res
    });
    return t;
}

pub fn part_two(input: &str) -> i32 {
    let t = input.lines().fold(0, |acc, l| {
        let mut s = l.replace("one", "one1one");
        s = s.replace("two", "two2two");
        s = s.replace("three", "three3three");
        s = s.replace("four", "four4four");
        s = s.replace("five", "five5five");
        s = s.replace("six", "six6six");
        s = s.replace("seven", "seven7seven");
        s = s.replace("eight", "eight8eight");
        s = s.replace("nine", "nine9nine");
        let first = s.chars().find(|ch| '0' <= *ch && *ch <= '9').unwrap();
        let last = s.chars().rev().find(|ch| '0' <= *ch && *ch <= '9').unwrap();
        let t = format!("{}{}", first, last);
        let res: i32 = t.parse().unwrap();
        acc + res
    });
    return t;
}

#[cfg(test)]
mod test {
    use super::{part_one, part_two};

    #[test]
    fn part_one_test() {
        let input = r#"1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet"#;
        let res = part_one(input);
        assert_eq!(res, 142);
    }

    #[test]
    fn parse_two_test() {
        let input = r#"two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen"#;
        let res = part_two(input);
        assert_eq!(res, 281);
    }
}

